builtin/v9/migration/miner.go

```diff
@@ -18,94 +24,102 @@ import (
 	}
 
-	newPrecommitOnChainInfos, err := adt.AsMap(wrappedStore, emptyMap, builtin.DefaultHamtBitwidth)
+	edv9, err := miner9.ConstructDeadline(ctxStore)
 	if err != nil {
-		return nil, xerrors.Errorf("failed to load empty map: %w", err)
+		return nil, xerrors.Errorf("failed to construct empty deadline v9: %w", err)
 	}
 
-	var info miner8.SectorPreCommitOnChainInfo
-	err = oldPrecommitOnChainInfos.ForEach(&info, func(key string) error {
-		var unsealedCid *cid.Cid
-		if len(info.Info.DealIDs) != 0 {
-			pieces := make([]abi.PieceInfo, len(info.Info.DealIDs))
-			for i, dealID := range info.Info.DealIDs {
-				deal, err := m.proposals.GetDealProposal(dealID)
-				if err != nil {
-					return xerrors.Errorf("error getting deal proposal: %w", err)
-				}
+	edv9cid, err := store.Put(ctx, edv9)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to put empty deadline v9: %w", err)
+	}
 
-				pieces[i] = abi.PieceInfo{
-					PieceCID: deal.PieceCID,
-					Size:     deal.PieceSize,
-				}
-			}
+	edsv9 := miner9.ConstructDeadlines(edv9cid)
+	edsv9cid, err := store.Put(ctx, edsv9)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct empty deadlines v9: %w", err)
 
-			commd, err := commp.GenerateUnsealedCID(info.Info.SealProof, pieces)
-			if err != nil {
-				return xerrors.Errorf("failed to generate unsealed CID: %w", err)
-			}
+	}
 
-			unsealedCid = &commd
-		}
+	return &minerMigrator{
+		emptyPrecommitOnChainInfosV9: emptyPrecommitMapCidV9,
+		emptyDeadlineV8:              edv8cid,
+		emptyDeadlinesV8:             edsv8cid,
+		emptyDeadlineV9:              edv9cid,
+		emptyDeadlinesV9:             edsv9cid,
+		proposals:                    marketProposals,
+		OutCodeCID:                   outCode,
+	}, nil
+}
 
-		err = newPrecommitOnChainInfos.Put(miner9.SectorKey(info.Info.SectorNumber), &miner9.SectorPreCommitOnChainInfo{
-			Info: miner9.SectorPreCommitInfo{
-				SealProof:     info.Info.SealProof,
-				SectorNumber:  info.Info.SectorNumber,
-				SealedCID:     info.Info.SealedCID,
-				SealRandEpoch: info.Info.SealRandEpoch,
-				DealIDs:       info.Info.DealIDs,
-				Expiration:    info.Info.Expiration,
-				UnsealedCid:   unsealedCid,
-			},
-			PreCommitDeposit: info.PreCommitDeposit,
-			PreCommitEpoch:   info.PreCommitEpoch,
-		})
+func (m minerMigrator) migratedCodeCID() cid.Cid {
+	return m.OutCodeCID
+}
 
-		if err != nil {
-			return xerrors.Errorf("failed to write new precommitinfo: %w", err)
-		}
+func (m minerMigrator) migrateState(ctx context.Context, store cbor.IpldStore, in actorMigrationInput) (*actorMigrationResult, error) {
+	var inState miner8.State
+	if err := store.Get(ctx, in.head, &inState); err != nil {
+		return nil, err
+	}
+	var inInfo miner8.MinerInfo
+	if err := store.Get(ctx, inState.Info, &inInfo); err != nil {
+		return nil, err
+	}
+	wrappedStore := adt8.WrapStore(ctx, store)
 
-		return nil
-	})
+	newPrecommits, err := m.migratePrecommits(ctx, wrappedStore, inState.PreCommittedSectors)
```

Comment:
Potentially worth caching

Reply:
Agreed. I'm gonna start an issue capturing possible further caching / speedups. I do want to keep the cache focused on highest impact stuff, because it can get big...